### PR TITLE
S7 register login enhancements

### PR DESCRIPTION
**Pull request by chorfiyoussef**
This PR introduces an enhancement of the registration workflow within the GoMeet app, encompassing multiple steps from username and email entry, password setup, personal information collection, and profile picture upload. Each step includes validations to ensure data integrity and user experience enhancements. The updates include:

- Username and Email Validation: Ensures uniqueness and formats correctness.
- Secure Password Input: Adds password confirmation and basic security checks.
- Personal Details Collection: Gathers essential user information like name, country, and contact.
- Profile Picture Upload: Offers a user-friendly interface to upload and preview profile images.

Additionally, existing UI tests for the registration process have been removed, as they are no longer relevant due to the significant changes in the workflow. New UI tests will be introduced in a subsequent PR to ensure high-quality code and functionality.

### CODE REVIEW

File: app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterScreen.kt [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+          Image(
+              painter = painterResource(id = R.drawable.gomeet_text),
+              contentDescription = "Go Meet",
+              modifier = Modifier.padding(top = 0.dp),
+              alignment = Alignment.Center,
+              colorFilter = ColorFilter.tint(MaterialTheme.colorScheme.tertiary))
+
+          Spacer(modifier = Modifier.height(screenHeight / 10))
+
+          fun nextState() {
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterScreen.kt:116]: This is a good idea, but upon going back a screen, the fields start out as empty again

File: app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterUsernameEmail.kt [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+        TextField(
+            value = username,
+            onValueChange = {
+              if (it.length < 26) {
+                charactersExceeded = false
+                username = it
+              } else {
+                charactersExceeded = true
+                username = it
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterUsernameEmail.kt:89]: Setting username here even if it's too long (even if you can't progress to the next screen) might introduce bugs, better to just set charactersExceeded to true in the else condition

File: app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterUsernameEmail.kt [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+            modifier = Modifier.fillMaxWidth())
+
+        if (charactersExceeded) {
+          Text(text = "Your Username Should Not Exceed 26 characters", color = Color.Red)
+        }
+
+        if (!firstClick && !isValidUsername) {
+          Text(text = "The Username is not valid or already taken", color = Color.Red)
+        }
+        Spacer(modifier = Modifier.size(16.dp))
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterUsernameEmail.kt:105]: Hardcoded spacer size again

**Review by DanThePol — COMMENTED** [review]
Great code, we won't need to refactor any of this later on. A few minor changes needed (see comments)

File: app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterPfp.kt [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+      rememberLauncherForActivityResult(
+          contract = ActivityResultContracts.GetContent(),
+          onResult = { uri: Uri? ->
+            pfpUri = uri
+            pfp = uri.toString()
+          })
+
+  Column(modifier = Modifier.fillMaxSize(), horizontalAlignment = Alignment.CenterHorizontally) {
+    Text(
+        text = if (pfp.isEmpty()) "Hi $name!\nWhat about a Profile Picture ?" else "You look good!",
```
[the remark this answers is not in the document]
Comment (chorfiyoussef) [reply app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterPfp.kt:69]: 😂😂😂

File: app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt [hunk 5]
```diff
@@ -85,6 +85,21 @@ class UserViewModel(userRepository: UserRepository) : ViewModel() {
     }
   }
 
+  /**
+   * Get all users of the app
+   *
+   * @return a list of all users
+   */
+  suspend fun getAllUser(): List<GoMeetUser>? {
```
[the remark this answers is not in the document]
Comment (chorfiyoussef) [reply app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt:93]: Fixed

File: app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterScreen.kt (hunk 1 above)
Comment (chorfiyoussef) [reply app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterScreen.kt:116]: Fixed !

File: app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterUsernameEmail.kt (hunk 2 above)
Comment (chorfiyoussef) [reply app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterUsernameEmail.kt:89]: In this case, the user will never exceed the maximum number of characters for a username (he will not be able to write more than 26 characters) and the `charactersExceeded` variable will be useless. I think we can leave like that to stay consistent with how other inputs are handled  (the user can write whatever he wants in the field, but the registration process is blocked if the input is not valid).

File: app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterUsernameEmail.kt (hunk 3 above)
Comment (chorfiyoussef) [reply app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterUsernameEmail.kt:105]: This is fixed

File: app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterNameCountryPhone.kt [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+                    Icons.AutoMirrored.Filled.ArrowForward,
+                    contentDescription = "Next",
+                    tint = DarkGrey,
+                    modifier = Modifier.size(60.dp))
+              }
+        }
+      }
+}
+
+fun getCountries(): ArrayList<String> {
```
[the remark this answers is not in the document]
Comment (chorfiyoussef) [reply app/src/main/java/com/github/se/gomeet/ui/authscreens/register/RegisterNameCountryPhone.kt:209]: done

**Review by DanThePol — APPROVED** [review]
Great PR, all minor issues mentioned during review have been fixed.

**Pull request merged**